query-translator/src/types.rs

```diff
@@ -0,0 +1,271 @@
+// Copyright 2016 Mozilla
+//
+// Licensed under the Apache License, Version 2.0 (the "License"); you may not use
+// this file except in compliance with the License. You may obtain a copy of the
+// License at http://www.apache.org/licenses/LICENSE-2.0
+// Unless required by applicable law or agreed to in writing, software distributed
+// under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
+// CONDITIONS OF ANY KIND, either express or implied. See the License for the
+// specific language governing permissions and limitations under the License.
+
+#![allow(dead_code, unused_imports)]
+
+use mentat_query_algebrizer::{
+    AlgebraicQuery,
+    ConjoiningClauses,
+    DatomsColumn,
+    DatomsTable,
+    QualifiedAlias,
+    SourceAlias,
+};
+
+use mentat_sql::{
+    BuildQueryError,
+    BuildQueryResult,
+    QueryBuilder,
+    QueryFragment,
+    SQLiteQueryBuilder,
+};
+
+//---------------------------------------------------------
+// A Mentat-focused representation of a SQL query.
+
+enum ColumnOrExpression {
+    Column(QualifiedAlias),
+    Integer(i64),             // Because it's so common.
+}
+
+type Name = String;
+struct Projection (ColumnOrExpression, Name);
+
+#[derive(Clone)]
+struct Op(String);      // TODO
+enum Constraint {
+    Infix {
+        op: Op,
+        left: ColumnOrExpression,
+        right: ColumnOrExpression
+    }
+}
+
+enum JoinOp {
+    Inner,
+}
+
+// Short-hand for a list of tables all inner-joined.
+struct TableList(Vec<SourceAlias>);
+
+struct Join {
+    left: TableOrSubquery,
+    op: JoinOp,
+    right: TableOrSubquery,
+    // TODO: constraints (ON, USING).
+}
+enum TableOrSubquery {
+    Table(SourceAlias),
+    // TODO: Subquery.
+}
+
+enum FromClause {
+    TableList(TableList),      // Short-hand for a pile of inner joins.
+    Join(Join),
+}
+
+struct SelectQuery {
+    projection: Vec<Projection>,
+    from: FromClause,
+    constraints: Vec<Constraint>,
+}
+
+//---------------------------------------------------------
+// Turn that representation into SQL.
+
+impl QueryFragment for ColumnOrExpression {
+    fn to_sql(&self, out: &mut QueryBuilder) -> BuildQueryResult {
+        use self::ColumnOrExpression::*;
+        match self {
+            &Column(QualifiedAlias(ref table, ref column)) => {
+                out.push_identifier(table.as_str())?;
```

Comment:
This is frustrating — I can't find a way to make the compiler accept anything that implements `AsRef<str>`. If I use `s: AsRef<str>` it complains because that's not sized. If I use `s: &AsRef<str>`, I can't pass `"foo"`. If I use `fn foo<T: AsRef<str>>(…, s: T)`, I get an error because you can't implement generic functions in traits.